Implemented presentation mathML printer

Subhash-Saurabh opened the pull request:
Implemented a class MathMLPresentationPrinter which allows printing presentation mathML code.

#### References to other Issues or PRs
 Fixes #11893

Subhash-Saurabh committed:
Implemented presentation mathML printer

mathml.py: Implemented a class MathMLPresentationPrinter
which allows printing presentation mathML code.
test_mathml.py: Added corresponding tests

Subhash-Saurabh committed:
Added Documentation changes and other Minor Fixes

Replaced MathMLPrinter with MathMLContentPrinter and MathMLPresentationPrinter
in test_pickling.py
Made Documentation Changes in printing.rst

Subhash-Saurabh committed:
Removed trailing whitespaces

Subhash-Saurabh commented:
@ylemkimon @asmeurer  can you please take a look at this PR.

ylemkimon commented on sympy/printing/mathml.py:
```diff
@@ -13,12 +13,10 @@
-class MathMLPrinter(Printer):
-    """Prints an expression to the MathML markup language
+class MathMLContentPrinter(Printer):
+    """Prints an expression to the Content MathML markup language.
```
I think it's better to create a base class which contains common codes in both printers, and make both printers inherit it.

ylemkimon replied:
Also, as they are public APIs, do not change existing names or create aliases from original names to them.

asmeurer replied:
I think the name change is appropriate, but you should add a `MathMLPrinter = MathMLContentPrinter` at the bottom of the file for backwards compatibility.

Subhash-Saurabh committed:
Added MathMLPrinter for backward compatibility

Subhash-Saurabh commented:
@asmeurer I made the changes you suggested. Is there anything else that I need to change.

asmeurer commented:
You should put the common code in a base class as @ylemkimon suggested. 

I would be nice if someone could review the MathML output here. 

How difficult would it be to take all the tested expressions and generate a simple HTML page that renders both the MathML, and the LaTeX (with MathJax) side-by-side, so that someone can easily compare them in a browser that supports MathML? That might require a bit of refactoring of the tests. 

Unless someone can suggest a better way to verify the tests. I suppose the alternative is for someone to verify everything by hand (I don't want to do it, but maybe another reviewer will).

Subhash-Saurabh committed:
Moved common code in class MathPrinter

Moved common code for both MathMLContentPrinter and MathMLPresentationPrinter
in new class MathPrinter.

ylemkimon commented on sympy/printing/mathml.py:
```diff
@@ -13,34 +13,89 @@
 
-class MathMLPrinter(Printer):
-    """Prints an expression to the MathML markup language
+class MathPrinter(Printer):
```
I think the name `MathPrinter` is too general and it should include 'mathml', like `MathMLPrinterBase`.

ylemkimon commented on sympy/printing/mathml.py:
```diff
@@ -13,34 +13,89 @@
-    """Prints an expression to the MathML markup language
+class MathPrinter(Printer):
+    # Contains common code required for MathMLContentPrinter and
+    # MathMLPresentationPrinter
```
Use triple quotes(docstring).

ylemkimon commented on sympy/printing/mathml.py:
```diff
@@ -13,34 +13,89 @@
-
-    References: https://www.w3.org/TR/MathML3/
-    """
     printmethod = "_mathml"
```
Define different `printmethod` for each printer.

Subhash-Saurabh replied:
I am thinking of renaming printmethods as `printmethod = "_mathml_content" ` and ` printmethod = "_mathml_presentation" ` . Will it be fine ?

ylemkimon commented on sympy/printing/mathml.py:
```diff
@@ -13,34 +13,89 @@
-    def __init__(self, settings=None):
-        Printer.__init__(self, settings)
-        from xml.dom.minidom import Document
-        self.dom = Document()
```
These codes can also go into the `MathPrinter` class.

ylemkimon commented on sympy/printing/mathml.py:
```diff
@@ -13,34 +13,89 @@
     def doprint(self, expr):
         """
-        Prints the expression as MathML.
+        Prints the expression as Content MathML.
```
This change seems to be no longer necessary.

ylemkimon commented on sympy/printing/mathml.py:
```diff
@@ -387,67 +442,485 @@ def _print_int(self, p):
+        self.dom = Document()
+
+        # Workaround to allow strings to remain unescaped
+        # Based on https://stackoverflow.com/questions/38015864/python-xml-dom-minidom-please-dont-escape-my-strings/38041194
```
Doesn't this affect `MathMLContentPrinter` too?

Subhash-Saurabh replied:
I don't think this affects `MathMLContentPrinter` as we don't need to add special characters (like integration and summation signs) ourselves. However we can add this for `MathMLContentPrinter` as well it won't make any difference.

Subhash-Saurabh committed:
Renamed MathPrinter and other requested changes

ylemkimon commented:
@asmeurer I think we could do an automated screenshot tests for LaTeX and MathML printing using selenium or similar.

Subhash-Saurabh commented:
@ylemkimon Is there anything that I can do for making the **verification of tests** easier ?

ylemkimon commented:
I'm currently working on SymPy Jupyter notebook extension, mainly consists of WYSIWYG editor. I'm also working on the testing infrastructure of it, and I think we can use it to test MathML printing.

ylemkimon commented on sympy/printing/mathml.py:
```diff
@@ -13,24 +13,37 @@
-    """Prints an expression to the MathML markup language
+class MathMLPrinterBase(Printer):
+    """Contains common code required for MathMLContentPrinter and
 
```
Unnecessary newline.

Subhash-Saurabh replied:
Should I remove newlines in other docstrings also like from `MathMLContentPrinter` and `MathMLPresentationPrinter`. I used newline because docstrings previously were written in this format .

ylemkimon replied:
@Subhash-Saurabh There is no need to put a newline within sentences. Between paragraphs and sections are okay.

ylemkimon reviewed (approved):
> Push it in now and improve upon it later.
(https://github.com/sympy/sympy/wiki/The-Zen-of-SymPy)

Subhash-Saurabh committed:
Removed newlines within sentences.
